### ISSUE TITLE
Issues with web-mode

### ISSUE THREAD

**Issue by LeartS** [alpaker/fill-column-indicator#46]
**Disclaimer:** not sure if a fill-column-indicator bug or a web-mode one.

I have a problem when editing files in web-mode with fci-mode active: there are operations that trigger a strange bug where the cursor goes to the fci-column.
For example:
- C-j at the end of a line creates a newline and positions the cursor to column 80 (or I guess whatever the set column is)
- auto tag closing (e.g. typing `<` when there is an open tag) closes the tag at column 80
- ...probably others.

```
Emacs 24.3.1
fill-column-indicator 20140302.1809
web-mode 20140413.1309
```

relevant configuration in init.el:

``` lisp
(require 'fill-column-indicator)
(setq fci-rule-column 80)
(add-hook 'after-change-major-mode-hook 'fci-mode)

(require 'web-mode)
(add-to-list 'auto-mode-alist '("\\.html$" . web-mode))
(defun my-web-mode-hook ()
  "Hooks for Web mode."
  (setq electric-indent-mode nil)
  (setq web-mode-markup-indent-offset 2)
  (setq web-mode-css-indent-offset 4)
  (setq web-mode-code-indent-offset 4)
  (setq web-mode-script-padding 2)
)
(add-hook 'web-mode-hook 'my-web-mode-hook)
```

![fci-bug](https://cloud.githubusercontent.com/assets/5588738/2713408/487c26c0-c4e6-11e3-9b60-aac629d09e56.png)

**Comment by peterhadlaw** [comment]
Hello, wondering if anyone was able to fix the indentation issues mentioned here, with Web Mode and FCI Mode?

**Comment by emmanueltouzery** [comment]
Unfortunately at least part of the issue is obviously still present as of today...

**Comment by alpaker** [comment]
To be honest, I haven't looked at this in months; I just haven't had much
time for hacking.   Sorry about that.

**Comment by emmanueltouzery** [comment]
it's obvious the issue isn't trivial...

I've configured my emacs like that for now to work around the issue:

```
  (add-hook 'after-change-major-mode-hook
            (lambda () (if (string= major-mode "web-mode")
                (turn-off-fci-mode) (turn-on-fci-mode))))
```

Fortunately HTML is a case where a 80-char width is much less relevant than in other file types.

**Comment by geraldus** [comment]
Hi friends!  I've just discovered `web-mode` and faced `fci-mode` issue also.  For now I simply disabled FCI when `web-mode` is active.  I really like FCI, but it conflicts with too many things and I still can't find good alternative.  By the way, `company-mode` also conflicts with FCI when showing completion pop-ups in similar manner, here is a workaround I'm using to resolve this issue:
```
(defvar-local company-fci-mode-on-p nil)

(defun company-turn-off-fci (&rest ignore)
  "Safely turn off Fill Column Indicator.
If `fci-mode' is enabled disable it and store its state in special variable.
Argument IGNORE is not used"
  (when (boundp 'fci-mode)
    (setq company-fci-mode-on-p fci-mode)
    (when fci-mode (fci-mode -1))))

(defun company-maybe-turn-on-fci (&rest ignore)
    "Turn on Fill Column Indicator if it was enabled.
If `fci-mode' was enabled turn it on.
Argument IGNORE is not used."
  (when company-fci-mode-on-p (fci-mode 1)))

(add-hook 'company-completion-started-hook 'company-turn-off-fci)
(add-hook 'company-completion-finished-hook 'company-maybe-turn-on-fci)
(add-hook 'company-completion-cancelled-hook 'company-maybe-turn-on-fci)
```
I think it is possible to exploit this solution somehow or maybe reordering buffer-change-hook functions  will solve this issue.  If I will find some neat easy workaround I will let you know.

Thank you for such great stuff.

**Comment by aspiers** [comment]
I'm seeing this too; I guess for now I'll also use the workaround to disable it with `web-mode`.

**Comment by dlight** [comment]
Yes, also here. Was the bug fixed in 2014, and later mysteriously returned?

This happens with other packages too (like haskell-mode for example). There ought to be a way to fix this without requiring cooperation from the affected packages.

**Comment by aspiers** [comment]
Yeah, actually fci-mode seems to be breaking almost everything for me right now :-(  20151030.1233 on 24.5.1.  I'll try updating it.

**Comment by dlight** [comment]
Just one thing, I'm discussing this issue with the web-mode author on fxbois/web-mode#242.

**Comment by zw963** [comment]
@fxbois , After twenty minutes test, finally, the anoise issue , which describe
in following content:

When I first paste one line content into `web-mode`. (C-y), and then
goto next line(C-n),  just pasted content wrong indent into the rightmost


It seem like caused by `Fill-Column-Indicator` mode, when close this mode
in `web-mode`,  it worked now.

Just for report.

**Comment by fxbois** [comment]
@zw963 I can provide help to the fill-column-indicator if needed

**Comment by orzechowskid** [comment]
instead of disabling fci-mode when web-mode is enabled, I turned the workaround above into advice and applied it to some web-mode functions (as well as the company-mode functions as described above and also mentioned [here](https://github.com/company-mode/company-mode/issues/180#issuecomment-55047120)):

```elisp
;; ~/.emacs

(defvar-local my-fci-mode-stack '()
  "track fci-mode state to aid advice functions.")

(defun fci-conditional-enable (&rest _)
  "Conditionally (re-)enable fci-mode."
  (when (eq (pop my-fci-mode-stack) t)
    (fci-mode t)))

(defun fci-get-and-disable (&rest _)
  "Store current status of fci-mode, and disable if needed."
  (when (boundp 'fci-mode)
    (push fci-mode my-fci-mode-stack)
    (when fci-mode
      (fci-mode -1))))

(defun fci-hack (advised-func &rest args)
  "Disable fci-mode, call ADVISED-FUNC with ARGS, then re-enable fci-mode."
  (progn
    (fci-get-and-disable)
    (apply advised-func args)
    (fci-conditional-enable)))

;; disable fci-mode while certain operations are being performed
(advice-add 'web-mode-on-after-change :around #'fci-hack)
(advice-add 'web-mode-on-post-command :around #'fci-hack)
(add-hook 'company-completion-started-hook 'fci-get-and-disable)
(add-hook 'company-completion-cancelled-hook 'fci-conditional-enable)
(add-hook 'company-completion-finished-hook 'fci-conditional-enable)
```
I don't know emacs all that well, but this appears to play nicely with how I use web-mode and company-mode.

**Comment by aspiers** [comment]
[**@emmanueltouzery**](https://github.com/emmanueltouzery) commented on [1 Jun 2015, 21:36 BST](https://github.com/alpaker/Fill-Column-Indicator/issues/46#issuecomment-107699526 "2015-06-01T20:36:07Z - Replied by Github Reply Comments"):
> I've configured my emacs like that for now to work around the issue:
> 
>       (add-hook 'after-change-major-mode-hook
>                 (lambda () (if (string= major-mode "web-mode")
>                     (turn-off-fci-mode) (turn-on-fci-mode))))

It took me a while to track this down as the culprit in my emacs setup, but unfortunately the above code is a bad idea, because it turns on fci-mode for *every* mode except `web-mode`, and that really doesn't make sense because it's broken in several other modes too - and outright unnecessary anyway.  For example it breaks `dired` where there is no good reason to have it enabled anyway. 

The solution is simple; just drop the else clause:

```
(add-hook 'after-change-major-mode-hook
          (lambda () (if (string= major-mode "web-mode")
              (turn-off-fci-mode))))
```

Hopefully that will teach me not to copy'n'paste code without first reading it ;-)

**Comment by alpaker** [comment]
Just FYI, I'm abandoning this package and welcoming requests to be take over as maintainer.

**Comment by alpaker** [comment]
The functionality provided by this package has now been implemented natively in Emacs as `display-fill-column-indicator-mode`, available as of version 27.0.90. Unless you are forced to use an older Emacs, you should use the native implementation instead of `fci-mode`. The former is unaffected by most of the compatibility issues that this add-on has.
